exams/api.py

```diff
@@ -32,3 +57,119 @@ def sso_digest(client_candidate_id, timestamp, session_timeout):
+    if not mmtrack.has_paid(course_run.edx_course_key):
+        errors_message = MESSAGE_NOT_ELIGIBLE_TEMPLATE.format(
+            user=mmtrack.user.username,
+            course_id=course_run.edx_course_key
+        )
+        raise ExamAuthorizationException(errors_message)
+
+    # if user paid for a course then create his exam profile if it is not created yet
+    ExamProfile.objects.get_or_create(profile=mmtrack.user.profile)
+
+    # if they didn't pass, they don't get authorized
+    if not mmtrack.has_passed_course(course_run.edx_course_key):
+        errors_message = MESSAGE_NOT_PASSED_OR_EXIST_TEMPLATE.format(
+            user=mmtrack.user.username,
+            course_id=course_run.edx_course_key
+        )
+        raise ExamAuthorizationException(errors_message)
+
+    ExamAuthorization.objects.get_or_create(
+        user=mmtrack.user,
+        course=course_run.course,
+        exam_run=exam_run,
+        defaults={  # write these out even though they're deprecated as putting nulls in here forfeits a db rollback
+            'date_first_eligible': exam_run.date_first_eligible,
+            'date_last_eligible': exam_run.date_last_eligible,
+        }
+    )
+    log.info(
+        '[Exam authorization] user "%s" is authorized for the exam for course id "%s"',
+        mmtrack.user.username,
+        course_run.edx_course_key
+    )
+
+
+def authorize_for_latest_passed_course(mmtrack, exam_run):
+    """
+    This walks the CachedEnrollments backwards chronologically and authorizes the first eligible one.
+
+    Args:
+        mmtrack (dashboard.utils.MMTrack): An instance of all user information about a program
+        exam_run (exams.models.ExamRun): the ExamRun to authorize the learner for
+    """
+    enrollments_qset = CachedEnrollment.objects.filter(
+        user=mmtrack.user,
+        course_run__course__id=exam_run.course_id,
+    ).order_by('-course_run__end_date')
+
+    if not enrollments_qset.exists():
+        return
+
+    for enrollment in enrollments_qset:
+        try:
+            authorize_for_exam_run(mmtrack, enrollment.course_run, exam_run)
+        except ExamAuthorizationException:
+            log.debug(
+                'Unable to authorize user: %s for exam on course_id: %s',
+                mmtrack.user.username,
+                enrollment.course_run.course.id
+            )
+        else:
+            break
+
+
+def bulk_authorize_for_exam_run(exam_run):
+    """
+    Authorize all eligible users for the given exam run
+
+    Args:
+        exam_run(exams.models.ExamRun): the exam run to authorize for
+    """
+    for program_enrollment in ProgramEnrollment.objects.filter(
+            program=exam_run.course.program
+    ).iterator():
+        mmtrack = get_mmtrack(
+            program_enrollment.user,
+            program_enrollment.program
+        )
+
+        try:
+            authorize_for_latest_passed_course(mmtrack, exam_run)
```

Comment:
Can you add a test which asserts that if the first call fails, the second call is still attempted?